File: AnomalyDetectorDLL/AnomalyDetectorDLL/anomaly_detection_util.cpp
/*
 * animaly_detection_util.cpp
 *
 * Author: Aviad Saadi, ID: 212330567
 */

#include "pch.h"
#include <math.h>
#include "anomaly_detection_util.h"

float avg(float* x, int size) {
    float sum = 0;
    for (int i = 0; i < size; ++i) {
        sum += x[i];
    }
    return sum / size;
}

// returns the variance of X and Y
float var(float* x, int size) {
    float* xSquare = new float[size];
    for (int i = 0; i < size; ++i) {
        xSquare[i] = pow(x[i], 2);
    }
    float res = avg(xSquare, size) - pow(avg(x, size), 2);
    delete[] xSquare;
    return  res;
}

// returns the covariance of X and Y
float cov(float* x, float* y, int size) {
    float* multiplyArray = new float[size];
    for (int i = 0; i < size; ++i) {
        multiplyArray[i] = x[i] * y[i];
    }
    float res = avg(multiplyArray, size) - avg(x, size) * avg(y, size);
    delete[] multiplyArray;
    return res;
}


// returns the Pearson correlation coefficient of X and Y
float pearson(float* x, float* y, int size) {
    return cov(x, y, size) / (sqrt(var(x, size) * var(y, size)));
}

// performs a linear regression and returns the line equation
Line linear_reg(Point** points, int size) {

    float* xArray = new float[size], * yArray = new float[size];
    for (int i = 0; i < size; ++i) {
        xArray[i] = points[i]->x;
        yArray[i] = points[i]->y;
    }
    float a = cov(xArray, yArray, size) / var(xArray, size);
    float b = avg(yArray, size) - a * avg(xArray, size);
    delete[] xArray;
    delete[] yArray;
    return Line(a, b);
}

// returns the deviation between point p and the line equation of the points
float dev(Point p, Point** points, int size) {
    Line line = linear_reg(points, size);
    return dev(p, line);
}

// returns the deviation between point p and the line
float dev(Point p, Line l) {
    float deviation = l.f(p.x) - p.y;
    if (deviation >= 0) {
        return deviation;
    }
    return -deviation;
}

File: FlightSimulator/FlightSimulator/AnomalyDetectionUtil.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace FlightSimulator
{
    public class AnomalyDetectionUtil
    {

        public class Line
        {
            private float a, b;
            public Line(float a, float b)
            {
                this.a = a;
                this.b = b;
            }

            public float f(float x)
            {
                return a * x + b;
            }

            public void setA(float value)
            {
                this.a = value;
            }

            public void setB(float value)
            {
                this.b = value;
            }

            public float getA()
            {
                return this.a;
            }

            public float getB()
            {
                return this.b;
            }
        };

        public class Point
        {
            
            private float x, y;
            public Point(float x, float y)
            {
                this.x = x;
                this.y = y;
            }

            public void setX(float value)
            {
                this.x = value;
            }

            public void setY(float value)
            {
                this.x = value;
            }

            public float getX()
            {
                return this.x;
            }

            public float getY()
            {
                return this.y;
            }
        };

        public float avg(List<float> x)
        {
            float sum = 0;
            for (int i = 0; i < x.Count(); ++i)
            {
                sum += x[i];
            }
            return sum / x.Count();
        }

        // returns the variance of X and Y
        public float var(List<float> x)
        {
            List<float> xSquare = new List<float>();
            for (int i = 0; i < x.Count(); ++i)
            {
                xSquare.Add(x[i] * x[i]);
            }
            return avg(xSquare) - (float)Math.Pow(avg(x), 2);
        }

        // returns the covariance of X and Y
        public float cov(List<float> x, List<float> y)
        {
            List<float> multiplyList = new List<float>();
            for (int i = 0; i < x.Count(); ++i)
            {
                multiplyList.Add(x[i] * y[i]);
            }
            return avg(multiplyList) - avg(x) * avg(y);
        }


        // returns the Pearson correlation coefficient of X and Y
        public float pearson(List<float> x, List<float> y)
        {
            return cov(x, y) / (float)(Math.Sqrt(var(x) * var(y)));
        }

        // performs a linear regression and returns the line equation
        public Line linear_reg(List<Point> points)
        {

            List<float> xArray = new List<float>();
            List<float> yArray = new List<float>();

            for (int i = 0; i < points.Count(); ++i)
            {
                xArray.Add(points[i].getX());
                yArray.Add(points[i].getY());
            }
            float a = cov(xArray, yArray) / var(xArray);
            float b = avg(yArray) - a * avg(xArray);
            return new Line(a, b);
        }

        // returns the deviation between point p and the line equation of the pointsLeft
        public float dev(Point p, List<Point> points)
        {
            Line line = linear_reg(points);
            return dev(p, line);
        }

        // returns the deviation between point p and the line
        public float dev(Point p, Line l)
        {
            float deviation = l.f(p.getX()) - p.getY();
            if (deviation >= 0)
            {
                return deviation;
            }
            return -deviation;
        }
    }
}
